File: packages/conformance-runner/src/wire.ts
/**
 * Copyright (c) 2026 OceanBase.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  validateOperationError,
  validateOperationRequest,
  validateOperationSuccess,
  validateWireJson,
  validateWireValue,
} from '@powercontext/protocol'
import type { CaseOutcome, ExpectedResult, WireCase } from './types.js'

function schemaValid(caseRow: WireCase): boolean {
  const schemaName = caseRow.schemaName
  if (schemaName === undefined) {
    return false
  }
  if (caseRow.rawJson !== undefined) {
    return validateWireJson(schemaName, caseRow.rawJson).valid
  }
  return validateWireValue(schemaName, caseRow.value).valid
}

function evaluateWire(caseRow: WireCase): boolean {
  if (caseRow.role === 'component') {
    return schemaValid(caseRow)
  }
  const operationId = caseRow.operationId
  if (operationId === undefined) {
    return false
  }
  if (caseRow.role === 'request') {
    if (caseRow.rawJson !== undefined && caseRow.schemaName !== undefined) {
      return validateWireJson(caseRow.schemaName, caseRow.rawJson).valid
    }
    return validateOperationRequest(operationId, caseRow.value).valid
  }
  if (caseRow.role === 'success') {
    return validateOperationSuccess(
      operationId,
      caseRow.status ?? 200,
      caseRow.contentType ?? 'application/json',
      caseRow.value,
    ).valid
  }
  return validateOperationError(operationId, caseRow.status ?? 422, caseRow.value).valid
}

export function runWireCase(
  caseRow: WireCase,
  expected: ExpectedResult | undefined,
): CaseOutcome {
  if (expected === undefined) {
    return { id: caseRow.id, status: 'fail', detail: 'missing expected result' }
  }
  const actual = evaluateWire(caseRow)
  if (actual !== expected.valid) {
    return {
      id: caseRow.id,
      status: 'fail',
      detail: `expected valid=${String(expected.valid)}, got ${String(actual)}`,
    }
  }
  return { id: caseRow.id, status: 'pass' }
}

File: conformance/runners/python/exporter/wire.py
# Copyright (c) 2026 OceanBase.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import Any

from .examples import example_for
from .openapi import component_schema
from .wire_case import wire_case
from .wire_mutations import invalid_object_cases


def request_cases(document: dict[str, Any], contract: dict[str, Any]) -> list[dict[str, Any]]:
    operation_id = contract["operationId"]
    schema_name = contract["request"]["schemaName"]
    if schema_name is None:
        return empty_request_cases(operation_id)
    schema = component_schema(document, schema_name)
    valid = example_for(document, schema)
    cases = [
        wire_case(
            f"{operation_id}.request.valid.minimal",
            operation_id=operation_id,
            schema_name=schema_name,
            role="request",
            expect="valid",
            tags=["minimal"],
            value=valid,
        )
    ]
    cases.extend(invalid_object_cases(document, operation_id, schema_name, schema, valid))
    return cases


def empty_request_cases(operation_id: str) -> list[dict[str, Any]]:
    return [
        wire_case(
            f"{operation_id}.request.valid.empty",
            operation_id=operation_id,
            role="request",
            expect="valid",
            tags=["empty"],
        ),
        wire_case(
            f"{operation_id}.request.invalid.extra",
            operation_id=operation_id,
            role="request",
            expect="invalid",
            tags=["extra"],
            value={},
        ),
    ]
